src/browser/ui/ReactDOMComponent.js

```diff
@@ -232,6 +234,24 @@ function processChildContext(context, tagName) {
   return context;
 }
 
+var onlyNestTextTags = {
+  'option': true
+};
+
+function getChildrenTextContent(tag, children) {
+  var childrenContent = '';
+  ReactChildren.forEach(children, function (c) {
+    if ("production" !== process.env.NODE_ENV) {
+      ("production" !== process.env.NODE_ENV ? warning(
+        !ReactElement.isValidElement(c),
+        tag + ' can not have nested tags'
+      ) : null);
+    }
+    childrenContent += c;
```

Comment:
https://github.com/facebook/react/blob/91194126d87c6165edf7d33ac67dd2fae86cf08f/src/utils/traverseAllChildren.js#L102

Children.each will skip boolean and undefined